akka-http-core/src/main/scala/akka/http/impl/engine/http2/Http2Multiplexer.scala

```diff
@@ -256,15 +257,15 @@ private[http2] trait Http2MultiplexerSupport { logic: GraphStageLogic with Stage
           else {
             val chosenId = prioritizer.chooseSubstream(sendableOutstreams.toSet)
             // expensive operation, to be optimized when prioritizers can be configured
-            // in 2.12.x there's no Queue.-=, when 2.12.x support is dropped, this can be
-            // changed to Queue.-=
-            sendableOutstreams.dequeueAll(_ == chosenId)
+            removeOutStream(chosenId)
             sendDataFrame(chosenId)
           }
 
-        def closeStream(streamId: Int): MultiplexerState =
-          // leave stream in sendableOutstreams, to be skipped in sendDataFrame
-          this
+        def closeStream(streamId: Int): MultiplexerState = {
+          // expensive operation, but only called for cancelled streams
+          removeOutStream(streamId)
```

Comment:
After reconsidering and having introduced two bugs here, I go with what @raboof suggested in https://github.com/akka/akka-http/pull/3873#discussion_r673799069 to take expensive operation here (only called for cancelled streams) and remove some of the complexity again.